[TwigBundle] Enable `#[AsTwigFilter]`, `#[AsTwigFunction]` and `#[AsTwigTest]` (#20879)
symfony/symfony-docs: templates.rst

    {# pass in the 3 optional arguments #}
    {{ product.price|price(2, ',', '.') }}

.. _templates-twig-filter-attribute:

> Create a class and fill in the logic::

Reviewer: ```suggestion
Create a class with a method that contains the filter logic, then add
the `#[AsTwigFilter]` attribute to define the name and options of
the Twig filter::
```

Author: Added, but somehow Github doesn't want me to resolve the conversation :/

Edit: nvm, Github not properly refreshing on my end

        }
    }

.. _templates-twig-function-attribute:

If you want to create a function instead of a filter, use the

> ``AsTwigFunction`` attribute::

Reviewer: ```suggestion
``#[AsTwigFunction]`` attribute::
```

    Along with custom filters and functions, you can also register
    `global variables`_.

.. versionadded:: 7.3

    Support for the ``#[AsTwigFilter]`` and ``#[AsTwigFunction]`` attributes was introduced in Symfony 7.3.

>     Previously, you had to use the ``getFilters()`` and ``getFunctions()`` methods.

Reviewer: ```suggestion
    Previously, you had to extend the `AbstractExtension` class, and override the 
    ``getFilters()`` and ``getFunctions()`` methods.
```

Register an Extension as a Service
..................................

> 

Reviewer: Just below, the tag to use when the class uses attributes is `twig.attribute_extension`. But this tag is added automatically by autoconfiguration.

Author: So do we mention that this tag is automatically added by autoconfiguration? Or not? I'm a bit scared it would just make the section a bit hard to read :/

Reviewer: Yes, you can mention that the tag is added by autoconfiguration when a Twig attribute is used on a method of the class. Otherwise, when autoconfiguration is not enabled, it needs to be added in the service definition.

    Along with custom filters and functions, you can also register
    `global variables`_.

.. versionadded:: 7.3

>     Support for the ``#[AsTwigFilter]`` and ``#[AsTwigFunction]`` attributes was introduced in Symfony 7.3.

Reviewer: Here you can add the `#[AsTwigTest]` to the list.

Including the code of the custom filters/functions in the Twig extension class
is the simplest way to create extensions. However, Twig must initialize all
extensions before rendering any template, even if the template doesn't use an

> extension. Note that if you use attributes, this part is not needed.

Reviewer: You could instead begin by mentioning the attributes in the very first sentence.

> When using attributes to extend Twig, the services are initialized only when the functions or filters are used to render the template. But in case you use the classic approach by extending the `AbstractExtension` class, Twig initialize all extensions before rendering...

Creating Lazy-Loaded Twig Extensions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> When using attributes to extend Twig, the services are initialized only when
> the functions or filters are used to render the template. But in case you use the
> classic approach by extending the ``AbstractExtension`` class, Twig initialize all extensions before
> rendering rendering any template, even if the template doesn't use an
> extension.

Reviewer: ```suggestion
When using attributes to extend Twig, the services are initialized only when
the functions or filters are used to render the template. But in case you use the
classic approach by extending the ``AbstractExtension`` class, Twig initializes all the extensions before
rendering any template, even if the extension is not used in the template.
```
